### PR TITLE
Basic where clause support

### PR DESCRIPTION
**Pull request by jussisaurio**
Sorry, I didn't realize @Feilkin had expressed interest in implementing this (#80) until I was mostly done... anyway, here's some kind of attempt. My bytecode/assembly-fu is quite basic 🍝

### CODE REVIEW

File: core/translate.rs [hunk 1]
```diff
@@ -272,11 +312,15 @@ fn translate_tables_begin(program: &mut ProgramBuilder, select: &mut Select) {
...
 
-fn translate_tables_end(program: &mut ProgramBuilder, select: &Select) {
+fn translate_tables_end(program: &mut ProgramBuilder, select: &Select) -> usize {
     // iterate in reverse order as we open cursors in order
+    let mut next_offset = std::usize::MAX;
     for table_loop in select.loops.iter().rev() {
         let cursor_id = table_loop.open_cursor;
         program.emit_insn(Insn::NextAsync { cursor_id });
+        if next_offset == std::usize::MAX {
+            next_offset = program.offset() - 1;
```
Comment (jussisaurio) [inline core/translate.rs:322]: this was sort of a yolo solution, can we always jump here when WHERE evaluates to false for a row.....?

**Comment by penberg** [comment]
@pereman2 Wanna review this?

**Comment by penberg** [comment]
@jussisaurio looks like limit 0 test is busted, can you please fix it?

**Comment by jussisaurio** [comment]
> @jussisaurio looks like limit 0 test is busted, can you please fix it?

Isn't it this test: `Test FAILED: 'select count(1) from users where 0;'` ? Anyway yeah, it returns `NULL` in this and `0` in sqlite, not sure why

Will take a look tomorrow

**Review by pereman2 — CHANGES_REQUESTED** [review]
Amazing job! 
Looks like you are generating more opcodes than necessary for simple (1) where clauses:
```sql
sqlite> explain select age from users where age < 3;
addr  opcode         p1    p2    p3    p4             p5  comment      
----  -------------  ----  ----  ----  -------------  --  -------------
0     Init           0     9     0                    0   Start at 9
1     OpenRead       0     2     0     10             0   root=2 iDb=0; users
2     Rewind         0     8     0                    0   
3       Column         0     9     1                    0   r[1]= cursor 0 column 9
4       Ge             2     7     1     BINARY-8       84  if r[1]>=r[2] goto 7
5       Column         0     9     3                    0   r[3]= cursor 0 column 9
6       ResultRow      3     1     0                    0   output=r[3]
7     Next           0     3     0                    1   
8     Halt           0     0     0                    0   
9     Transaction    0     0     2     0              1   usesStmtJournal=0
10    Integer        3     2     0                    0   r[2]=3
11    Goto           0     1     0                    0   
```

your explain:

```sql
> explain select sum(age) from users where age < 3;
addr  opcode             p1    p2    p3    p4             p5  comment
----  -----------------  ----  ----  ----  -------------  --  -------
0     Init               0     19    0                    0   Start at 19
1     OpenReadAsync      0     2     0                    0   root=2
2     OpenReadAwait      0     0     0                    0   
3     RewindAsync        0     0     0                    0   
4     RewindAwait        0     16    0                    0   
5       Column           0     9     1                    0   r[1]= cursor 0 column 9
6       Integer          2     3     0                    0   
7       Lt               1     2     10                   0   r[1] < r[2] -> 10
8       Integer          0     0     0                    0   
9       Goto             0     11    0                    0   
10      Integer          0     1     0                    0   
11      IfNot            0     14    0                    0   r[0] -> 14
12      Column           0     9     4                    0   r[4]= cursor 0 column 9
13      AggStep          0     4     3     sum            0   accum=r[3] step(4)
14    NextAsync          0     0     0                    0   
15    NextAwait          0     4     0                    0   
16    AggFinal           0     3     0     sum            0   accum=r[3]
17    ResultRow          3     1     0                    0   output=r[3..4]
18    Halt               0     0     0                    0   
19    Transaction        0     0     0                    0   
20    Goto               0     1     0                    0   
```

this can be simplfied by generating an opcode checks the opposite and jumps to end of loop like sqlite.

In the future we will want this inside the table_begin loop generation as right now, columns from outer loops are generated in inner loops which is intended badly optimized for now.  As a reference:
```sql
sqlite> explain select age, price from users, products where age < 3 and  price > 2;
addr  opcode         p1    p2    p3    p4             p5  comment      
----  -------------  ----  ----  ----  -------------  --  -------------
0     Init           0     17    0                    0   Start at 17
1     OpenRead       0     2     0     10             0   root=2 iDb=0; users
2     OpenRead       1     3     0     3              0   root=3 iDb=0; products
3     Rewind         0     16    0                    0   
4       Column         0     9     1                    0   r[1]= cursor 0 column 9
5       Ge             2     15    1     BINARY-8       84  if r[1]>=r[2] goto 15
6       Rewind         1     16    0                    0   
7         Column         1     2     1                    0   r[1]= cursor 1 column 2
8         RealAffinity   1     0     0                    0   
9         Le             3     14    1     BINARY-8       85  if r[1]<=r[3] goto 14
10        Column         0     9     4                    0   r[4]= cursor 0 column 9
11        Column         1     2     5                    0   r[5]= cursor 1 column 2
12        RealAffinity   5     0     0                    0   
13        ResultRow      4     2     0                    0   output=r[4..5]
14      Next           1     7     0                    1   
15    Next           0     4     0                    1   
16    Halt           0     0     0                    0   
17    Transaction    0     0     2     0              1   usesStmtJournal=0
18    Integer        3     2     0                    0   r[2]=3
19    Integer        2     3     0                    0   r[3]=2
20    Goto           0     1     0                    0   
```
there column for first table is generated in line 8, which we don't do currently, but as you can see there is a `LE` there which indicates this will be generated inside table begin. If possible I would move insert_where_clause to table begin so that @penberg suggest the name change of the function ;)

**Comment by penberg** [comment]
@jussisaurio @pereman2 I think once we have the test passing, let's merge this to avoid conflicts. But totally agree on moving the call. Codegen optimizations also make sense

**Comment by penberg** [comment]
The test itself is just wrong. I suspect you got bitten by the fact that our aggregates was bit broken before but got fixed in 93aefa3ee6b8f5a30a5286c28558556387f8ac4c:

The following fixes the problem:

```
diff --git a/testing/all.test b/testing/all.test
index 0ab665e..252ae03 100755
--- a/testing/all.test
+++ b/testing/all.test
@@ -117,7 +117,7 @@ do_execsql_test where-clause-unary-true {
 # not correct? should be 0?
 do_execsql_test where-clause-unary-false {
     select count(1) from users where 0;
-} {NULL}
+} {0}

 do_execsql_test where-clause-no-table-unary-true {
     select 1 where 1;
```

Merging...

**Pull request merged**